frontend/src/metabase/visualizations/echarts/pie/option/index.ts

```diff
@@ -0,0 +1,222 @@
+) {
+  const graphicOption = cloneDeep(TOTAL_GRAPHIC_OPTION);
+
+  // Don't display any text if there isn't enough width
+  if (outerRadius * 2 < DIMENSIONS.totalDiameterThreshold) {
+    return graphicOption;
+  }
+
+  graphicOption.children.forEach(child => {
+    child.style.fontFamily = renderingContext.fontFamily;
+  });
+
+  graphicOption.children[0].style.text = formatters.formatMetric(
+    hoveredIndex != null
+      ? chartModel.slices[hoveredIndex].data.displayValue
+      : chartModel.total,
+  );
+  graphicOption.children[0].style.fill = renderingContext.getColor("text-dark");
+
+  graphicOption.children[1].style.text =
+    hoveredIndex != null
+      ? formatters
+          .formatDimension(chartModel.slices[hoveredIndex].data.key)
+          .toUpperCase()
+      : TOTAL_TEXT;
+  graphicOption.children[1].style.fill =
+    renderingContext.getColor("text-light");
+
+  return graphicOption;
+}
+
+function getRadiusOption(sideLength: number) {
+  const outerRadius = sideLength / 2;
+  const innerRadius = outerRadius * DIMENSIONS.slice.innerRadiusRatio;
+
+  return { outerRadius, innerRadius };
+}
+
+function getIsLabelVisible(
+  label: string,
+  slice: d3.layout.pie.Arc<PieSliceData>,
+  innerRadius: number,
+  outerRadius: number,
+  fontSize: number,
+  renderingContext: RenderingContext,
+) {
+  // We use the law of cosines to determine the length of the chord with the
+  // same endpoints as the arc. The label should be shorter than this chord, and
+  // it should be shorter than the donutWidth.
+  //
+  // See the following document for a more detailed explanation:
+  // https://www.notion.so/metabase/Pie-Chart-Label-Visibility-Explanation-4cf366a78c6a419d95763a431a36b175?pvs=4
+  let arcAngle = slice.startAngle - slice.endAngle;
+  arcAngle = Math.min(Math.abs(arcAngle), Math.PI - 0.001);
+
+  const innerCircleChordLength = Math.sqrt(
+    2 * innerRadius * innerRadius -
+      2 * innerRadius * innerRadius * Math.cos(arcAngle),
+  );
+  const donutWidth = outerRadius - innerRadius;
+  const maxLabelDimension = Math.min(innerCircleChordLength, donutWidth);
+
+  const fontStyle = {
+    size: fontSize,
+    family: renderingContext.fontFamily,
+    weight: DIMENSIONS.slice.label.fontWeight,
+  };
+  const labelWidth = renderingContext.measureText(label, fontStyle);
+  const labelHeight = renderingContext.measureTextHeight(label, fontStyle);
+
+  return (
+    labelWidth + DIMENSIONS.slice.label.padding <= maxLabelDimension &&
+    labelHeight + DIMENSIONS.slice.label.padding <= maxLabelDimension
+  );
+}
+
+export function getPieChartOption(
+  chartModel: PieChartModel,
+  formatters: PieChartFormatters,
+  settings: ComputedVisualizationSettings,
+  renderingContext: RenderingContext,
+  sideLength: number,
+  hoveredIndex?: number,
+): EChartsOption {
+  // Sizing
+  const seriesOption = cloneDeep(SUNBURST_SERIES_OPTION); // deep clone to avoid sharing assigned properties with other instances
+  if (!seriesOption.label) {
```

Comment:
With the current approach `label` cannot be missing

Reply:
That's fine, `label` will always be defined.